pkg/snet/sudph/client.go

```diff
@@ -0,0 +1,428 @@
+
+	hs := directtransport.ResponderHandshake(func(f2 directtransport.Frame2) error {
+		c.mx.Lock()
+		defer c.mx.Unlock()
+
+		var ok bool
+		if lis, ok = c.lMap[f2.DstAddr.Port]; !ok {
+			return errors.New("not listening on given port")
+		}
+
+		return nil
+	})
+
+	connConfig := directtransport.ConnConfig{
+		Log:       c.log,
+		Conn:      udpConn,
+		LocalPK:   c.lPK,
+		LocalSK:   c.lSK,
+		Deadline:  time.Now().Add(directtransport.HandshakeTimeout),
+		Handshake: hs,
+		FreePort:  nil,
+		Encrypt:   true,
+		Initiator: false,
+	}
+
+	conn, err := directtransport.NewConn(connConfig)
+	if err != nil {
+		return err
+	}
+
+	return lis.Introduce(conn)
+}
+
+// Dial dials a new sudph.Conn to specified remote public key and port.
+func (c *Client) Dial(ctx context.Context, rPK cipher.PubKey, rPort uint16) (*directtransport.Conn, error) {
+	if c.isClosed() {
+		return nil, io.ErrClosedPipe
+	}
+
+	c.log.Infof("Dialing PK %v", rPK)
+
+	visorData, err := c.addressResolver.ResolveSUDPH(ctx, rPK)
+	if err != nil {
+		return nil, fmt.Errorf("resolve PK (holepunch): %w", err)
+	}
+
+	c.log.Infof("Resolved PK %v to visor data %v, dialing", rPK, visorData)
+
+	udpConn, err := c.dialVisor(visorData)
+	if err != nil {
+		return nil, err
+	}
+
+	c.log.Infof("Dialed %v:%v@%v", rPK, rPort, udpConn.RemoteAddr())
+
+	lPort, freePort, err := c.p.ReserveEphemeral(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("ReserveEphemeral: %w", err)
+	}
+
+	hs := directtransport.InitiatorHandshake(c.lSK, dmsg.Addr{PK: c.lPK, Port: lPort}, dmsg.Addr{PK: rPK, Port: rPort})
+
+	connConfig := directtransport.ConnConfig{
+		Log:       c.log,
+		Conn:      udpConn,
+		LocalPK:   c.lPK,
+		LocalSK:   c.lSK,
+		Deadline:  time.Now().Add(directtransport.HandshakeTimeout),
+		Handshake: hs,
+		FreePort:  freePort,
+		Encrypt:   true,
+		Initiator: true,
+	}
+
+	sudpConn, err := directtransport.NewConn(connConfig)
+	if err != nil {
+		return nil, fmt.Errorf("newConn: %w", err)
+	}
+
+	return sudpConn, nil
+}
+
+// Listen creates a new listener for sudp hole punch.
+// The created Listener cannot actually accept remote connections unless Serve is called beforehand.
+func (c *Client) Listen(lPort uint16) (*directtransport.Listener, error) {
+	if c.isClosed() {
+		return nil, io.ErrClosedPipe
+	}
+
+	ok, freePort := c.p.Reserve(lPort)
+	if !ok {
+		return nil, errors.New("port is already occupied")
+	}
+
+	c.mx.Lock()
+	defer c.mx.Unlock()
+
+	lAddr := dmsg.Addr{PK: c.lPK, Port: lPort}
+	lis := directtransport.NewListener(lAddr, freePort)
+	c.lMap[lPort] = lis
+
+	return lis, nil
+}
+
+// Close closes the Client.
+func (c *Client) Close() error {
+	if c == nil {
+		return nil
+	}
+
+	c.once.Do(func() {
+		close(c.done)
+
+		c.mx.Lock()
+		defer c.mx.Unlock()
+
+		if err := c.addressResolver.Close(); err != nil {
+			c.log.WithError(err).Warnf("Failed to close address resolver client")
+		}
+
+		for _, lis := range c.lMap {
+			_ = lis.Close() // nolint:errcheck
```

Comment:
Why won't we log errors?

Reply:
I copied this code from `stcp` and did no changes to it because I was going to combine all transports in one. I agree that we should have an error check here, but would you mind if this would be done in a PR for #409? This code won't be duplicated 6 times there.

Reply:
Of course, do as comfortable